[AOTInductor] Remove call to aot_autograd when receiving ExportedProgram

angelayi opened the pull request:
https://github.com/pytorch/pytorch/issues/105555

Existing flow first exports and then calls torch._inductor.aot_compile. However, export calls aot_autograd with the core aten decomposition table, and then torch._inductor.aot_compile calls aot_autograd again with the inductor decomposition table. The 2nd calling of aot_autograd is supposedly causing some problems, and seems excessive, so instead we will create a new function, torch._export.aot_compiler which will export using the inductor decomposition table, pass it to inductor's compile_fx_aot, and because it has already been exported, avoid recalling aot_autograd.

```
def aot_compile(
    f: Callable,
    args: Tuple[Any],
    kwargs: Optional[Dict[str, Any]] = None,
    constraints: Optional[List[Constraint]] = None,
) -> Tuple[str, ExportedProgram]:
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

angelayi commented on torch/_export/_experimental.py:
```diff
@@ -0,0 +1,42 @@
+
+__all__ = ["aot_compile"]
+
+def aot_compile(
```
We can't have torch._inductor.aot_compile take in an ExportedProgram because of circular import errors...
But another solution is that because we're in python we could also just ignore the annotations and pass torch._inductor.aot_compile an ExportedProgram and import + check inside that function.

desertfire commented on torch/_inductor/config.py:
```diff
@@ -63,6 +63,10 @@
 
+# If true, specifies that we are running AOTInductor from an exported program,
+# implying that aot_autograd has already been run.
+aot_from_export = False
```
I would prefer to make AOTInductor takes exported program as a mandatory requirement, with some checking in comple_fx_aot. wrapper_call in AotCodeCache is too hacky and not reliable anyway, and I think giving user too much flexibility as in the old way isn't necessarily a good choice.

desertfire commented on torch/_export/__init__.py:
```diff
@@ -145,12 +143,18 @@ def export(
     kwargs = kwargs or {}
 
+    if decomposition_table is None:
+        decomposition_table = core_aten_decompositions()
```
Do we want to keep this only for AOTInductor? Also when decomposition_table is not None, you want to merge the two dicts.

desertfire replied:
`dict.update` can take precedence, but I guess arguably either way is fine. Just need to make the behavior clear in the user API comment/doc.

angelayi replied:
Hmm, but what if a user removed something from the core aten decomp table? Then by adding it, it'll get decomposed again.

zhxchen17 replied:
I guess from API design point of view you can do two things as Bin suggested:
1. overwrite
2. merge
overwrite is fine but the downside is user is implicitly depending on the fact that export() uses core aten IR by default, which is a little bit leaky abstraction to me, but I guess in practice it's fine?

If we wanna go with merge route, then we need an "addition set" and a "minus set" which are two arguments.

angelayi replied:
My preference is to go with overwriting, but another idea is, what if we use a context manager to specify the decomp table rather having it as an argument to export? 

My feeling is that the manipulation of the decomp table will be used by super users, so if they do want to specify their own decompositions then they can specify it through a context manager like:
```
with set_export_decomp_table(inductor_decomp):
    ep = export(...)
```

zhxchen17 replied:
I think it's still leaky whether you make it context manager or not. User just express their intention differently here:
1. For overwrite, user basically says: no matter what export() does, use MY decomp table.
2. For merge, use says: on top of export() decomp behavior, I want these to be decomped and those to be not decomped. 

I think 1 and 2 are not the same although in practice they could achieve the same thing. I think context manager v.s. argument is just implicit vs explicit over an API which is orthogonal to the semantics difference I mentioned above.

angelayi replied:
Chatted offline: since we don't have a strong use case for passing in the decomp table yet, we will just patch the decomp table directly in our helper function

desertfire commented on torch/_inductor/compile_fx.py:
```diff
@@ -217,14 +220,16 @@ def materialize(x):
                     ]
-                ]
+                else:
+                    real_inputs = [materialize(x) for x in V.real_inputs]
```
This simple materialize was causing problem for some models previously. Do you think the inputs will all be real tensors in the export+aot_compile path? If yes, we should be fine here.

angelayi replied:
I think the original code also expects the inputs to always be real tensors? The original code differs by just getting the parameters from the tracing context, but from export the parameters are already passed in as inputs.

desertfire commented:
Please also remember to update https://github.com/pytorch/pytorch/blob/78fffe890627bc020c9bd82a27a19d042f31695e/benchmarks/dynamo/common.py#L1090

angelayi commented:
After elias lands his diff, I can rebase, import to fbsource, and fix internal use cases.

desertfire commented on test/inductor/test_aot_inductor.py:
```diff
@@ -28,14 +29,13 @@ def load(cls, model, example_inputs, example_outputs):
+        exported = torch._export.export(
+            model,
+            example_inputs,
+            decomposition_table=fast_random_decomps(),
```
`select_decomp_table()` may be better as it respects `config.fallback_random`.

desertfire commented on benchmarks/dynamo/common.py:
```diff
@@ -1110,13 +1110,8 @@ def load(cls, model, example_inputs, eager_forward):
 
             # The exact API is subject to change
             exported = torch._export.export(model, example_inputs)
```
pass in `decomposition_table`?

desertfire commented on torch/_inductor/graph.py:
```diff
@@ -938,6 +938,9 @@ def compile_to_fn(self):
             code, linemap = self.codegen()
             output_code_log.debug("Output code: \n%s", code)
 
+            if config.cpp_wrapper:
```
This will always be true because of `if self.aot_mode and self.cpp_wrapper`? I think can just get rid of `AotCodeCache.compile`.

desertfire reviewed (approved):
CI failure is related to this file, https://github.com/pytorch/pytorch/blob/main/test/cpp/aot_inductor/test.py, which needs to be updated.

Approve to unblock, need to fix small issues before landing.

desertfire commented on benchmarks/dynamo/common.py:
```diff
@@ -1109,14 +1110,11 @@ def load(cls, model, example_inputs, eager_forward):
-            flat_example_inputs = fx_pytree.tree_flatten_spec(
-                example_inputs, exported.call_spec.in_spec
+            exported = torch._export.export(
+                model, example_inputs, decomposition_table=select_decomp_table()
```
Also we probably want to provide a specific API to target Export+AOTInductor. It's awkward for most end users to figure out which decomp table to pass in here.

desertfire replied:
Ok for this PR. We can address it later.

angelayi replied:
Makes sense -- do you have an opinion on where this API should live? In export or in inductor?

angelayi replied:
Added the API to torch._export

zhxchen17 commented on torch/_export/__init__.py:
```diff
@@ -119,14 +119,15 @@ class ExportDynamoConfig:
     constraints: Optional[List[Constraint]] = None,
+    *,
+    decomposition_table: Optional[Dict] = None,
+    _add_runtime_assertions=True,
+    _functionalize_runtime_assertions=False,
```
remove these?

angelayi replied:
Oh, somehow they were there...

zhxchen17 commented:
export() part has some major nits need to be addressed, otherwise looks fine overall

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -888,11 +888,8 @@ def compile(cls, graph, source_code, cuda):
-            return cls.cache[key], *(None for i in range(len(graph.graph_outputs) - 1))
-
-        return wrapper_call
+            return cls.cache[key]
+        return None
```
Why do we need these changes ?

angelayi replied:
Instead of returning a function, @desertfire said we should just directly return the so_path, so this just gets rid of the wrapper.

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -217,18 +220,20 @@ def materialize(x):
-                            if param is not None
-                        ],
-                        *V.real_inputs,
+                if torch._guards.TracingContext.get():
```
could you add a comment what's happening here ?

also, this is my bad, but while you're here would you mind refomating the 
```
                    real_inputs = [
                        materialize(x)
                        for x in [
                            *[
                                param
                                for param in torch._guards.TracingContext.get().params_flat
                                if param is not None
                            ],
                            *V.real_inputs,
                        ]
                    ]
```

lines

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -531,6 +536,9 @@ def fx_codegen_and_compile(
                         context.output_strides.append(None)
             compiled_fn = graph.compile_to_fn()
 
+            if _in_aot_compilation:
+                return compiled_fn
```
just for my own understanding, why do we need this ?

angelayi replied:
My understanding is, if we're just aot compiling, we do not need to construct the followup `CompiledFXGraph` and we can just return the "compiled_fn" which in this case is just the so_path from AOTCodeCache

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -1079,6 +1087,10 @@ def bw_compiler(model: torch.fx.GraphModule, example_inputs):
     )
+    if _in_aot_compilation:
+        with V.set_fake_mode(fake_mode), compiled_autograd.disable():
+            return fw_compiler(model_, example_inputs_)
```
This is skipping freezing because we changed it to invoking `fw_compiler` instead of `inference_compiler`

angelayi replied:
I have a followup PR to change this to `inference_compiler`: https://github.com/pytorch/pytorch/pull/106239, but right now freezing doesn't work with the c++ wrapper so I wasn't sure if I should make that change here.

eellison replied:
sounds good

eellison reviewed (changes requested):
we're not invoking freezing currently

kimishpatel commented:
Angela, can you say, and perhaps add to summary, what this PR is doing?

angelayi committed:
[AOTInductor] Remove call to aot_autograd when receiving ExportedProgram (#105977)

Summary:
https://github.com/pytorch/pytorch/issues/105555

Existing flow first exports and then calls torch._inductor.aot_compile. However, export calls aot_autograd with the core aten decomposition table, and then torch._inductor.aot_compile calls aot_autograd again with the inductor decomposition table. The 2nd calling of aot_autograd is supposedly causing some problems, and seems excessive, so instead we will create a new function, torch._export.aot_compiler which will export using the inductor decomposition table, pass it to inductor's compile_fx_aot, and because it has already been exported, avoid recalling aot_autograd.

```
def aot_compile(
    f: Callable,
    args: Tuple[Any],
    kwargs: Optional[Dict[str, Any]] = None,
    constraints: Optional[List[Constraint]] = None,
) -> Tuple[str, ExportedProgram]:
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

Pull Request resolved: https://github.com/pytorch/pytorch/pull/105977

Reviewed By: eellison

Differential Revision: D47998435

Pulled By: angelayi

fbshipit-source-id: 5efbaacf731ec5a2e5c496807c3d15dbdd5160f9
